preview/Win7Msix/Win7MSIXInstaller/ValidateTargetDeviceFamily.cpp

```diff
@@ -0,0 +1,174 @@
+    ComPtr<IMsixElement> element;
+    RETURN_IF_FAILED(domElement->GetDocumentElement(&element));
+
+    /// Obtain the TargetDeviceFamily tag info
+    ComPtr<IMsixElementEnumerator> dependencyEnum;
+    RETURN_IF_FAILED(element->GetElements(
+        L"/*[local-name()='Package']/*[local-name()='Dependencies']/*[local-name()='TargetDeviceFamily']",
+        &dependencyEnum));
+
+    BOOL hc = FALSE;
+    RETURN_IF_FAILED(dependencyEnum->GetHasCurrent(&hc));
+    if (!hc)
+    {
+        TraceLoggingWrite(g_MsixTraceLoggingProvider,
+            "No Target device family Found",
+            TraceLoggingLevel(WINEVENT_LEVEL_ERROR));
+        return HRESULT_FROM_WIN32(ERROR_INSTALL_REJECTED);
+    }
+
+    ComPtr<IMsixElement> dependencyElement;
+    RETURN_IF_FAILED(dependencyEnum->GetCurrent(&dependencyElement));
+
+    Text<wchar_t> targetDeviceFamilyName;
+    RETURN_IF_FAILED(dependencyElement->GetAttributeValue(L"Name", &targetDeviceFamilyName));
+    m_targetDeviceFamilyName = targetDeviceFamilyName.Get();
+
+    Text<wchar_t> minVersion;
+    RETURN_IF_FAILED(dependencyElement->GetAttributeValue(L"MinVersion", &minVersion));
+    std::wstring manifestMinVersion = minVersion.Get();
+
+    /// Major version
+    size_t start = 0;
+    size_t end = manifestMinVersion.find_first_of(L'.');
+    m_majorVersion = manifestMinVersion.substr(start, end - start);
+
+    /// Minor version
+    manifestMinVersion.replace(start, end - start + 1, L"");
+    end = manifestMinVersion.find_first_of(L'.');
+    m_minorVersion = manifestMinVersion.substr(start, end - start);
+
+    /// Build number
+    manifestMinVersion.replace(start, end - start + 1, L"");
+    end = manifestMinVersion.find_first_of(L'.');
+    m_buildNumber = manifestMinVersion.substr(start, end - start);
+
+    return S_OK;
+}
+
+bool ValidateTargetDeviceFamily::IsTargetDeviceFamilyNameCompatibleWithOS()
+{
+    if (isWindowsProductTypeDesktop()) /// Desktop OS
+    {
+        if (m_targetDeviceFamilyName == L"MsixCore.Server" || m_targetDeviceFamilyName == L"MsixCore.Desktop")
+        {
+            return true;
+        }
+    }
+    else if (isWindowsProductTypeServer()) /// Server
+    {
+        if (m_targetDeviceFamilyName == L"MsixCore.Server")
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool ValidateTargetDeviceFamily::IsManifestVersionCompatibleWithOS()
+{
+    OSVERSIONINFOEX osvi;
+    DWORDLONG dwlConditionMask = 0;
+    int op = VER_GREATER_EQUAL;
+
+    ZeroMemory(&osvi, sizeof(OSVERSIONINFOEX));
+    osvi.dwOSVersionInfoSize = sizeof(OSVERSIONINFOEX);
+    osvi.dwMajorVersion = std::stoi(m_majorVersion.c_str());
+    osvi.dwMinorVersion = std::stoi(m_minorVersion.c_str());
+    osvi.dwBuildNumber = std::stoi(m_buildNumber.c_str());
+
+    if (isWindowsProductTypeDesktop())
+    {
+        osvi.wProductType = VER_NT_WORKSTATION;
+    }
+    else if (isWindowsProductTypeServer())
+    {
+        osvi.wProductType = VER_NT_SERVER;
+    }

```

Comment:
remove this. You did it in IsTargetDeviceFamilyNameCompatibleWithOS()

Reply:
Removed.